firedrake/slate/slate.py

```diff
@@ -0,0 +1,616 @@
+
+from firedrake.utils import cached_property
+
+from ufl.form import Form
+from ufl.algorithms.map_integrands import map_integrand_dags
+from ufl.algorithms.multifunction import MultiFunction
+from ufl.domain import join_domains, sort_domains
+
+
+__all__ = ['TensorBase', 'Tensor', 'Inverse', 'Transpose',
+           'UnaryOp', 'Negative', 'BinaryOp', 'TensorAdd',
+           'TensorSub', 'TensorMul']
+
+
+class TensorBase(object):
+    """An abstract SLATE node class.
+
+    .. warning::
+
+       Do not instantiate this class on its own. This is an abstract
+       node class; is not meant to be worked with directly. Only use
+       the appropriate subclasses.
+    """
+
+    id = 0
+
+    def __init__(self):
+        """Constructor for the TensorBase abstract class."""
+        self.id = TensorBase.id
+        TensorBase.id += 1
+        self._integral_domains = None
+        self._subdomain_data = None
+        self._hash = None
+
+    @cached_property
+    def shape(self):
+        """Computes the shape information of the local tensor."""
+        shape = []
+        for i, arg in enumerate(self.arguments()):
+            V = arg.function_space()
+            shapelist = []
+            for fs in V:
+                shapelist.append(fs.fiat_element.space_dimension() * fs.dim)
+            shape.append(sum(shapelist))
+        return tuple(shape)
+
+    @cached_property
+    def shapes(self):
+        """Computers the internal shape information of its components.
+        This is particularly useful to know if the tensor comes from a
+        mixed form.
+        """
+        shapes = {}
+        for i, arg in enumerate(self.arguments()):
+            V = arg.function_space()
+            shapelist = []
+            for fs in V:
+                shapelist.append(fs.fiat_element.space_dimension() * fs.dim)
+            shapes[i] = tuple(shapelist)
+        return shapes
+
+    @cached_property
+    def rank(self):
+        """Returns the rank information of the tensor object."""
+        return len(self.arguments())
+
+    @property
+    def inv(self):
+        return Inverse(self)
+
+    @property
+    def T(self):
+        return Transpose(self)
+
+    def __add__(self, other):
+        return TensorAdd(self, other)
+
+    def __radd__(self, other):
+        """Ordering of tensor addition does not matter."""
+        return self.__add__(other)
+
+    def __sub__(self, other):
+        return TensorSub(self, other)
+
+    def __rsub__(self, other):
+        """Ordering of tensor subtraction does not matter."""
+        return self.__sub__(other)
+
+    def __mul__(self, other):
+        return TensorMul(self, other)
+
+    def __rmul__(self, other):
+        """Tensor multiplication is not commutative in general."""
+        return TensorMul(other, self)
+
+    def __neg__(self):
+        return Negative(self)
+
+    def __pos__(self):
+        return self
+
+    def check_integrals(self, integrals):
```

Comment:
Shouldn't be here: generic nodes don't contain integrals, this should move to `Tensor`.